compiler-core/src/type_/expression.rs

```diff
@@ -4428,8 +4458,45 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
         value
     }
 
+    /// Given a function being called and one of its arguments, this returns the
+    /// argument's definition location if it was defined by that same function.
+    /// For example:
+    ///
+    /// ```gleam
+    /// fn wibble(x) {
+    /// //        ^ Defined by the wibble function
+    ///   wibble(x, 1)
+    /// //       ^ And used in the recursive call unchanged!
+    /// }
+    /// ```
+    ///
+    fn argument_used_in_recursive_call(
+        &self,
+        called_function: &TypedExpr,
+        argument: &TypedExpr,
+        argument_index: usize,
+    ) -> Option<SrcSpan> {
+        let argument = argument.var_constructor()?;
+        let (function_defining_argument, index_in_argument_list) =
+            argument.variant.argument_function_and_index()?;
+
+        let (module, function) = called_function
+            .var_constructor()
+            .and_then(|constructor| constructor.variant.function_module_and_name())?;
+
+        if self.environment.current_module == module
+            && function == function_defining_argument
+            && argument_index == index_in_argument_list
+        {
+            Some(argument.definition_location().span)
+        } else {
+            None
+        }
```

Comment:
It's a bit unsatisfying to me that this data is all being fetched on every call when it will have already been fetched just before during analysis.

Reply:
Sorry I don't see where this data is fetched. Here we're checking for the specific case of a function being passed an argument it defines and so they both need to have that specific shape that is not checked elsewhere when typing them.

Reply:
What do you mean by have a specific shape?

Reply:
We don't store the current function calls' arguments, but that it very easy to add.

It is all the other information that is being re-fetched here.